sdk/modelsrepository/Azure.IoT.ModelsRepository/src/MetadataScheduler.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+
+namespace Azure.IoT.ModelsRepository
+{
+    /// <summary>
+    /// MetadataScheduler tracks when metadata should be fetched from a repository.
+    /// </summary>
+    internal class MetadataScheduler
+    {
+        private DateTime _lastFetched;
+        private readonly TimeSpan _desiredElapsedTimeSpan;
+        private bool _initialFetch;
+
+        /// <param name="desiredElapsedTimeSpan">A desired <c>TimeSpan</c> that is used to calculate elapsed time periods.</param>
+        public MetadataScheduler(TimeSpan desiredElapsedTimeSpan)
+        {
+            _lastFetched = DateTime.MinValue;
+            _desiredElapsedTimeSpan = desiredElapsedTimeSpan;
+            _initialFetch = true;
+        }
+
+        /// <summary>
+        /// Updates the last fetched DateTime attribute of the scheduler to <c>DateTime.Now</c>.
+        /// </summary>
+        public void Reset()
+        {
+            if (_initialFetch)
+            {
+                _initialFetch = false;
+            }
+
+            _lastFetched = DateTime.Now;
+        }
+
+        /// <summary>
+        /// Determines whether the desired time span has elapsed with respect to the last fetched DateTime attribute.
+        /// </summary>
+        public bool HasElapsed()
```

Comment:
so if `Reset()` is not called, `HasElapsed()` will always return true? I guess I am a bit confused about the use of `_initialFetch`. Can you elaborate?

Reply:
The idea for `_initialFetch` is to support an initial metadata state in the client instance. `HasElapsed()` indicates whether a fresh metadata fetch should occur. `Reset()` is meant to be called after a fetch attempt has occurred. Once an initial fetch has occurred, the `_desiredElapsedTimeSpan` is used to calculate when a subsequent metadata fetch/refresh should occur.